lib/improver/utilities/statistical_operations.py

```diff
@@ -0,0 +1,213 @@
+        """
+        probabilities = self.create_probability_cube(reference_cube)
+
+        array_shape = list(cube.shape)
+        array_shape.insert(0, 2)
+        array_shape = tuple(array_shape)
+        percentile_bounds = np.full(array_shape, -1, dtype=float)
+        height_bounds = np.full(array_shape, -1001., dtype=float)
+        height_bounds[1] = -1.
+
+        percentile_coordinate, = [coord for coord in reference_cube.coords()
+                                  if 'percentiles' in coord.name()]
+        percentiles = percentile_coordinate.points
+
+        for index, pslice in enumerate(reference_cube.slices_over(
+                percentile_coordinate)):
+            indices = (cube.data < pslice.data if self.inverse_ordering else
+                       cube.data > pslice.data)
+            percentile_bounds[0, indices] = percentiles[index]
+            height_bounds[0, indices] = pslice.data[indices]
+            try:
+                # Usual behaviour where the orography falls between heights
+                # corresponding to percentiles.
+                percentile_bounds[1, indices] = percentiles[index + 1]
+                height_bounds[1, indices] = reference_cube[index+1].data[
+                    indices]
+            except IndexError:
+                # Invoked if we have reached the top of the available heights.
+                percentile_bounds[1, indices] = percentiles[index]
+                height_bounds[1, indices] = pslice.data[indices]
+
+        with np.errstate(divide='ignore'):
+            interpolants, = ((cube.data - height_bounds[0]) /
+                             np.diff(height_bounds, n=1, axis=0))
+
+        with np.errstate(invalid='ignore'):
+            probabilities.data, = (percentile_bounds[0] + interpolants *
+                                   np.diff(percentile_bounds, n=1, axis=0))
+        probabilities.data = probabilities.data/100.
+
+        above_top_band = np.isinf(interpolants)
+        below_bottom_band = height_bounds[0] < -1000
+        probabilities.data[below_bottom_band] = 0.
+        probabilities.data[above_top_band] = 1.
+
+        return probabilities
+
+    def process(self, cube):
+        """
+        Slice the cube over realization if present and call the percentile
+        interpolation method.
+
+        Args:
+            cube (iris.cube.Cube):
+                A cube of values, that effectively behave as thresholds, for
+                which it is desired to obtain probability values from a
+                percentiled reference cube.
+        Returns:
+            output_cubes (iris.cube.Cube):
+                A cube of probabilities obtained by interpolating the
+                percentile values.
+        """
+        try:
+            self.reference_cube.coord_dims('realization')
+        except CoordinateNotFoundError:
+            cube_slices = [self.reference_cube]
+        else:
+            cube_slices = self.reference_cube.slices_over('realization')
```

Comment:
Why only slicing over realization? What if the percentiles have other coords e.g. Time?

Reply:
I'm not sure if this is based on some assumption regarding the input cube. This doesn't seem ideal.